crates/prover/src/rpc.rs

```diff
@@ -0,0 +1,288 @@
+use citrea_common::utils::{
+    check_l2_range_exists, extract_sequencer_commitments, filter_out_proven_commitments,
+};
+use jsonrpsee::core::RpcResult;
+use jsonrpsee::proc_macros::rpc;
+use jsonrpsee::types::error::{INTERNAL_ERROR_CODE, INTERNAL_ERROR_MSG};
+use jsonrpsee::types::ErrorObjectOwned;
+use sov_db::ledger_db::ProverLedgerOps;
+use sov_db::schema::types::BatchNumber;
+use sov_modules_core::{Spec, Storage};
+use sov_rollup_interface::da::{BlobReaderTrait, BlockHeaderTrait, DaSpec, SequencerCommitment};
+use sov_rollup_interface::services::da::{DaService, SlotData};
+use sov_rollup_interface::zk::StateTransitionData;
+use tokio::sync::Mutex;
+use tracing::{debug, error};
+
+use crate::da_block_handler::get_state_transition_data_from_commitments;
+
+pub(crate) struct RpcContext<C, Da, DB>
+where
+    C: sov_modules_api::Context,
+    Da: DaService,
+    DB: ProverLedgerOps,
+{
+    pub da_service: Arc<Da>,
+    pub ledger: DB,
+    pub sequencer_da_pub_key: Vec<u8>,
+    pub sequencer_pub_key: Vec<u8>,
+    pub l1_block_cache: Arc<Mutex<L1BlockCache<Da>>>,
+    pub phantom: PhantomData<fn() -> C>,
+}
+
+#[rpc(client, server)]
+pub trait ProverRpc {
+    /// Generate state transition data for the given L1 block height, and return the data as a borsh serialized hex string.
+    #[method(name = "prover_generateInput")]
+    async fn generate_input(&self, l1_height: u64) -> RpcResult<String>;
+}
+
+pub struct ProverRpcServerImpl<C, Da, DB>
+where
+    C: sov_modules_api::Context,
+    Da: DaService,
+    DB: ProverLedgerOps + Send + Sync + 'static,
+{
+    context: Arc<RpcContext<C, Da, DB>>,
+}
+
+impl<C, Da, DB> ProverRpcServerImpl<C, Da, DB>
+where
+    C: sov_modules_api::Context,
+    Da: DaService,
+    DB: ProverLedgerOps + Send + Sync + 'static,
+{
+    pub fn new(context: RpcContext<C, Da, DB>) -> Self {
+        Self {
+            context: Arc::new(context),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<C: sov_modules_api::Context, Da: DaService, DB: ProverLedgerOps + Send + Sync + 'static>
+    ProverRpcServer for ProverRpcServerImpl<C, Da, DB>
+{
+    async fn generate_input(&self, l1_height: u64) -> RpcResult<String> {
+        debug!("Prover: prover_generateInput");
+
+        let l1_block: <Da as DaService>::FilteredBlock = self
+            .context
+            .da_service
+            .get_block_at(l1_height)
+            .await
+            .map_err(|e| {
+                ErrorObjectOwned::owned(
+                    INTERNAL_ERROR_CODE,
+                    INTERNAL_ERROR_MSG,
+                    Some(format!("{e}",)),
+                )
+            })?;
+
+        let mut da_data: Vec<<<Da as DaService>::Spec as DaSpec>::BlobTransaction> =
+            self.context.da_service.extract_relevant_blobs(&l1_block);
+
+        // if we don't do this, the zk circuit can't read the sequencer commitments
+        da_data.iter_mut().for_each(|blob| {
+            blob.full_data();
+        });
```

Comment:
We're already doing this inside `extract_sequencer_commitments. Isn't this redundant?